### PR TITLE
feat(GUI): add electron-native file-picker component

### PR DESCRIPTION
**Pull request by Shou**
We add a file-picker written with Rendition/React. It is activated with
the `ETCHER_EXPERIMENTAL_FILE_PICKER` environment variable. Further
customisation can be done with the `ETCHER_FILE_BROWSER_CONSTRAIN_FOLDER`
variable that takes a path and allows one to constrain the file-picker to
a folder.

Related: https://github.com/resin-io/etcher/issues/2238
Related: https://github.com/resin-io/etcher/issues/2285
Change-Type: patch
Changelog-Entry: Add electron-native file-picker component.

### CODE REVIEW

**Comment by Shou** [comment]
image

**Comment by Shou** [comment]
Filters will be in a follow-up PR.

File: lib/gui/app/components/file-selector/styles/_file-selector.scss [hunk 1]
```diff
@@ -0,0 +1,9 @@
+
+.modal-file-selector-modal {
+  min-width: 98vw;
+
+  > .modal-content {
+    min-height: 95vh;
+    max-height: 95vh;
```
Comment (Shou) [inline lib/gui/app/components/file-selector/styles/_file-selector.scss]: I should make these `calc(100vh - 10px)` like the Wifi modal.

File: lib/gui/app/models/storage.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
+
```
Comment (Shou) [inline lib/gui/app/models/storage.js]: This needs to be rebased on https://github.com/resin-io/etcher/pull/2276 and https://github.com/resin-io/etcher/pull/2311

File: lib/gui/css/main.css [hunk 3]
```diff
@@ -6487,6 +6487,12 @@ svg-icon {
   max-height: 200px;
   overflow-y: auto; }
 
+.modal-file-selector-modal {
+  min-width: 98vw; }
+  .modal-file-selector-modal > .modal-content {
+    min-height: 95vh;
+    max-height: 95vh; }
```
Comment (Shou) [inline lib/gui/css/main.css]: These should also be `calc(100vh - 10px)` like the Wifi modal, and ideally only specify these in one location like the Wifi modal if possible.

File: lib/gui/app/components/file-selector/file-selector/file-selector.jsx [hunk 4]
```diff
@@ -0,0 +1,591 @@
...
+ */
+const MAX_DIR_CRUMBS = 3
+
+/*
+ * TODO
+ * - move state to redux store, e.g. highlighted file
+ * - make back button
+ * - filters with no hidden files
+ * - 
+ */
```
Comment (Shou) [inline lib/gui/app/components/file-selector/file-selector/file-selector.jsx]: Remove these

File: lib/gui/app/components/file-selector/file-selector/file-selector.jsx [hunk 5]
```diff
@@ -0,0 +1,591 @@
...
+ *
+ * @param {String} fullpath - path string
+ * @returns {Array<Object>} - all subpaths as file objects
+ *
+ * @example
+ * const subpaths = subpaths('/home/user/Downloads')
+ * console.log(subpaths.map(file => file.fullpath))
+ * > [ '/', '/home', '/home/user', '/home/user/Downloads' ]
+ */
+const subpaths = (fullpath) => {
```
Comment (Shou) [inline lib/gui/app/components/file-selector/file-selector/file-selector.jsx]: Move this to `lib/shared/files`. Also pass labels to the functions so we can label generic folders and not just root.

File: lib/gui/app/components/file-selector/file-selector/file-selector.jsx [hunk 6]
```diff
@@ -0,0 +1,591 @@
...
+})`
+  color: ${props => props.color};
+  font-size: ${props => props.size};
+`
+
+const FileLink = styled((props) => {
+  const icon = props.isDirectory ? 'faFolder' : 'faFileAlt'
+
+  return (
+    <div className={ props.className }
```
Comment (Shou) [inline lib/gui/app/components/file-selector/file-selector/file-selector.jsx]: Use Flex element here

File: lib/gui/app/components/file-selector/file-selector/file-selector.jsx [hunk 7]
```diff
@@ -0,0 +1,591 @@
...
+  max-height: 128px;
+  margin: 0 10px 10px 10px;
+  background-color: ${ props => props.highlight ? colors.secondary.background : 'none' };
+  transition: 0.15s background-color ease-in-out;
+  color: ${ colors.primary.color };
+  cursor: pointer;
+  border-radius: 2px;
+`
+
+const Header = styled.header`
```
Comment (Shou) [inline lib/gui/app/components/file-selector/file-selector/file-selector.jsx]: We can do `styled(Flex)` here

File: lib/gui/app/components/file-selector/file-selector/file-selector.jsx [hunk 8]
```diff
@@ -0,0 +1,591 @@
...
+  flex: 0 0 auto;
+  align-items: baseline;
+  margin: 10px 15px;
+
+  > * {
+    margin: 5px;
+  }
+`
+
+const Main = styled.main`
```
Comment (Shou) [inline lib/gui/app/components/file-selector/file-selector/file-selector.jsx]: And here

File: lib/gui/app/components/file-selector/file-selector/file-selector.jsx [hunk 9]
```diff
@@ -0,0 +1,591 @@
...
+  display: flex;
+  flex: 1;
+  margin-left: 20px;
+`
+
+const Footer = styled.footer`
```
Comment (Shou) [inline lib/gui/app/components/file-selector/file-selector/file-selector.jsx]: And here!

File: lib/gui/app/components/file-selector/file-selector/file-selector.jsx [hunk 10]
```diff
@@ -0,0 +1,591 @@
...
+  justify-content: flex-end;
+  margin: 10px 20px;
+  flex: 0 0 auto;
+
+  > * {
+    margin: 0 10px;
+  }
+`
+
+const Left = styled.div`
```
Comment (Shou) [inline lib/gui/app/components/file-selector/file-selector/file-selector.jsx]: We don't need this, `Flex` can replace this entirely.

File: lib/gui/app/components/file-selector/file-selector/file-selector.jsx [hunk 11]
```diff
@@ -0,0 +1,591 @@
...
+  display: flex;
+  flex: 0 0 auto;
+`
+
+const Right = styled.div`
```
Comment (Shou) [inline lib/gui/app/components/file-selector/file-selector/file-selector.jsx]: We don't need this, `Flex` can replace this entirely.

File: lib/gui/app/components/file-selector/file-selector/file-selector.jsx [hunk 12]
```diff
@@ -0,0 +1,591 @@
...
+    margin-bottom: 15px;
+  }
+
+  > button {
+    margin-bottom: 10px;
+    text-align: start;
+  }
+`
+
+const labels = {
```
Comment (Shou) [inline lib/gui/app/components/file-selector/file-selector/file-selector.jsx]: Remove this, see above comment.

**Comment by jhermsmeier** [comment]
@Shou does this support multiple mountpoints (i.e. if the source plugged in has multiple partitions)?

File: tests/gui/pages/main.spec.js [hunk 13]
```diff
@@ -35,6 +35,10 @@ require.extensions['.html'] = (module, filename) => {
   })
 }
 
+// NOTE(Shou): since we don't test React yet we just ignore JSX files
+// eslint-disable-next-line node/no-deprecated-api
+require.extensions['.jsx'] = _.constant(null)
```
Comment (jhermsmeier) [inline tests/gui/pages/main.spec.js]: Why is this necessary?

Comment (Shou) [reply tests/gui/pages/main.spec.js]: Tests explode on JSX files because they aren't handled by Webpack.

File: lib/gui/app/components/file-selector/file-selector/file-selector.jsx [hunk 14]
```diff
@@ -0,0 +1,620 @@
...
+
+            this.setState({ error: invalidImageError })
+            analytics.logEvent('Invalid image', image)
+            return
+          }
+
+          return Bluebird.try(() => {
+            let message = null
+
+            if (supportedFormats.looksLikeWindowsImage(image.path)) {
```
Comment (lurch) [inline lib/gui/app/components/file-selector/file-selector/file-selector.jsx]: Is this duplicating logic that you already have elsewhere?
